File: cs_version5/cs_version5/Technical_manager.cs
/***********************************************************************
 * Module:  Technical_manager.cs
 * Author:  Anton
 * Purpose: Definition of the Class Technical_manager
 ***********************************************************************/

using System;
using System.Collections.Generic;

public class Technical_manager : Worker
{
    public Technical_manager()
    {
        test = new List<Test>(1);
        interview = new List<Interview>(1);
        test.Add(new Test());
        interview.Add(new Interview());
        //name = "";
        requirementLevel = 1;
        commutabilityLevel = 1;
        Console.WriteLine("Technical_manager was created (default)");
    }
	public Technical_manager(string nam, int rLevel, int cLevel) : base(nam)
    {
        test.Add(new Test());
        interview.Add(new Interview());
        //name = nam;
        requirementLevel = rLevel;
        commutabilityLevel = cLevel;
        Console.WriteLine("Technical_manager was created (inicialisation)");
    }
	public Technical_manager(Technical_manager sTechnical_manager)
    {
        test = sTechnical_manager.test;
        interview = sTechnical_manager.interview;
        //name = sTechnical_manager.name;
        requirementLevel = sTechnical_manager.requirementLevel;
        commutabilityLevel = sTechnical_manager.commutabilityLevel;
        Console.WriteLine("Technical_manager was created (copy)");
    }
   
   public string getName()
   {
       return name;
   }
   public int getReqLevel()
   {
       return requirementLevel;
   }
   public int getComLevel()
   {
       return commutabilityLevel;
   }
   public string getState()
   {
       return state;
   }
   
    public void setState(string s)
    {
	    state = s;
    }

    public void prepareTest(List<string> list)
    {
        int numOfQuations = 0;
        for (int i = 0; i < list.Count; i++)
        {
            if (test.Count == 0 || i > 0)
            {
                test.Add(new Test());
            }
            test[test.Count - 1].setType(list[i]);
            Console.WriteLine("Enter number of quastions");
            numOfQuations = int.Parse(Console.ReadLine());
            List<string> lis = new List<string>(1);
            for (int j = 0; j < numOfQuations; j++)
            {
                string str;
                Console.WriteLine("Enter dificulty of quastion");
                str = Console.ReadLine();
                lis.Add(str);
            }
            test[test.Count - 1].setQuestions(lis);
        }
    }

	public string checkTest(List<string> listOfAnswers)
    {
        int amountOfcorrectAnsw = 0;
	    for (int i = 0; i < listOfAnswers.Count; i++)
	    {
		    amountOfcorrectAnsw = listOfAnswers[i] == "done" ? amountOfcorrectAnsw + 1 : amountOfcorrectAnsw;
	    }
	    if ((float)amountOfcorrectAnsw / (float)listOfAnswers.Count >= 0.66)
	    {
		    return "DoneTest";
	    }
	    else
	    {
		    return "FailTest";
	    }
    }

	public string holdingInterview(int cLevel, int wLevel, string name)
    {
        if (interview.Count == 0 || !(interview[interview.Count - 1].getWhoPass().Length == 0))
	    {
		    interview.Add(new Interview());
	    }
	    bool res = false;
	    res = (cLevel * commutabilityLevel) / (wLevel * requirementLevel) != 0 ? true : false;
	    interview[interview.Count - 1].setResult(res);
	    interview[interview.Count - 1].setWhoPass(name);
	    if (res == true)
	    {
		    return "getWork";
	    }
	    else
	    {
		    return "Fail";
	    }
    }

	public Test getTestForCandidate(string competence)
    {
        for (int i = 0; i < test.Count; i++)
	    {
		    if (test[i].getType() == competence)
		    {
			    return test[i];
		    }
	    }
        return test[0];
    }

   public List <Interview> interview;
   
   public List<Test> test;
  

   private string name;
   private int requirementLevel;
   private int commutabilityLevel;
   private string state;

}

File: version_3/version_3/Technical_manager.cpp
/***********************************************************************
* Module:  Technical_manager.cpp
* Author:  Anton
* Modified: Monday, November 7, 2016 7:09:04 PM
* Purpose: Implementation of the class Technical_manager
***********************************************************************/
#include <vector>
#include <string>
#include "Interview.h"
#include "Test.h"
#include "Technical_manager.h"

//constructors
Technical_manager::Technical_manager() :
	test(1, new Test),
	interview(1, new Interview),
	name("name"),
	requirementLevel(1),
	commutabilityLevel(1),
	state("free")
{
	/*for (int i = 0; i < 1; i++)
	{
	interview[i] = new Interview;
	}
	for (int i = 0; i < 1; i++)
	{
	test[i] = new Test;
	}*/
	//std::cout << "Technical_manager was created (default)" << std::endl;
}

Technical_manager::Technical_manager(std::string name, int rLevel, int cLevel) :
	test(1, new Test),
	interview(1, new Interview),
	name(name),
	requirementLevel(rLevel),
	commutabilityLevel(cLevel),
	state("free")
{
	/*for (int i = 0; i < 1; i++)
	{
	interview[i] = new Interview;
	}
	for (int i = 0; i < 1; i++)
	{
	test[i] = new Test;
	}*/
	//std::cout << "Technical_manager was created (inicialisation)" << std::endl;
}

Technical_manager::Technical_manager(Technical_manager &sTechnical_manager) :
	test(1, new Test),
	interview(1, new Interview),
	name(sTechnical_manager.name),
	requirementLevel(sTechnical_manager.requirementLevel),
	commutabilityLevel(sTechnical_manager.commutabilityLevel),
	state(sTechnical_manager.state)
{
	//std::cout << "Technical_manager was created (copy)" << std::endl;
}

Technical_manager::Technical_manager(std::string str) :
	test(1, new Test),
	interview(1, new Interview),
	name(str.substr(0, 5)),
	requirementLevel(atoi(str.substr(5, 2).c_str())),
	commutabilityLevel(atoi(str.substr(7, 2).c_str())),
	state("free")
{
	//name = ;
	//requirementLevel = ;
	//commutabilityLevel = ;
	//std::cout << "Technical_manager was created (transform)" << std::endl;
}

//destructor
Technical_manager::~Technical_manager()
{
	//std::cout << "Technical manager was destroyed" << std::endl;
	//std::cout << std::endl;
}

////////////////////////////////////////////////////////////////////////
// Name:       Technical_manager::prepareTest()
// Purpose:    Implementation of Technical_manager::prepareTest()
// Return:     void
////////////////////////////////////////////////////////////////////////

void Technical_manager::prepareTest(std::vector<std::string> list)
{
	int numOfQuations = 0;
	for (int i = 0; i < list.size(); i++)
	{
		if (test.empty() || i > 0)
		{
			test.push_back(new Test);
		}
		test[test.size() - 1]->setType(list[i]);
		std::cout << "Enter number of quastions" << std::endl;
		std::cin >> numOfQuations;
		std::vector<std::string> list;
		for (int j = 0; j < numOfQuations; j++)
		{
			std::string str;
			std::cout << "Enter dificulty of quastion" << std::endl;
			std::cin >> str;
			list.push_back(str);
		}
		test[test.size() - 1]->setQuestions(list);//TODO: get index of test with nacessary type
	}
	// TODO : implement
}

////////////////////////////////////////////////////////////////////////
// Name:       Technical_manager::checkTest()
// Purpose:    Implementation of Technical_manager::checkTest()
// Return:     void
////////////////////////////////////////////////////////////////////////

std::string Technical_manager::checkTest(std::vector<std::string> listOfAnswers)
{
	// TODO : implement
	int amountOfcorrectAnsw = 0;
	for (int i = 0; i < listOfAnswers.size(); i++)
	{
		amountOfcorrectAnsw = listOfAnswers[i] == "done" ? amountOfcorrectAnsw + 1 : amountOfcorrectAnsw;
	}
	if ((float)amountOfcorrectAnsw / (float)listOfAnswers.size() >= 0.66)
	{
		return "DoneTest";
	}
	else
	{
		return "FailTest";
	}
}

////////////////////////////////////////////////////////////////////////
// Name:       Technical_manager::holdingInterview()
// Purpose:    Implementation of Technical_manager::holdingInterview()
// Return:     void
////////////////////////////////////////////////////////////////////////

std::string Technical_manager::holdingInterview(int cLevel, int wLevel, std::string name)//REMAKE!!!
{
	// TODO : implement
	if (interview.empty() || !interview[interview.size() - 1]->getWhoPass().empty())
	{
		interview.push_back(new Interview);
	}
	bool res = false;
	res = (cLevel * commutabilityLevel) / (wLevel * requirementLevel) != 0 ? true : false;
	interview[interview.size() - 1]->setResult(res);
	interview[interview.size() - 1]->setWhoPass(name);
	if (res == true)
	{
		return "getWork";
	}
	else
	{
		return "Fail";
	}
}

Test* Technical_manager::getTestForCandidate(std::string competence)
{
	for (int i = 0; i < test.size(); i++)
	{
		if (test[i]->getType() == competence)
		{
			return test[i];
		}
	}
}

////////////////////////////////////////////////////////////////////////
// Name:       Technical_manager::get()
// Purpose:    Implementation of Technical_manager::get()
// Return:     void
////////////////////////////////////////////////////////////////////////

//getters
std::string Technical_manager::getName(void)
{
	return name;
}

int Technical_manager::getReqLevel(void)
{
	return requirementLevel;
}

int Technical_manager::getComLevel(void)
{
	return commutabilityLevel;
}

std::string Technical_manager::getState(void)
{
	return state;
}

////////////////////////////////////////////////////////////////////////
// Name:       Technical_manager::set()
// Purpose:    Implementation of Technical_manager::set()
// Return:     void
////////////////////////////////////////////////////////////////////////

//setters
void Technical_manager::setName(std::string n)
{
	name = n;
}

void Technical_manager::setReqLevel(int r)
{
	requirementLevel = r;
}

void Technical_manager::setComLevel(int c)
{
	commutabilityLevel = c;
}

void Technical_manager::setState(std::string s)
{
	state = s;
}